### PR TITLE
Parameter variable substitution support for conditionals

### PR DESCRIPTION
**Pull request by dibyom**
# Changes

Add parameter variable substitution support for conditionals 
A Condition's `spec.Check` can refer to its declared parameters using
the ${params.param-name} syntax just like in a Task. Values can be
provided via the pipelinerun's parameters or via default values in the
condition or in the PipelineTaskCondition
# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

Part of #1019, #1137 

# Release Notes

```
- Adds the ability to declare and use parameters in a Condition 
whose values can be defined in a pipelinerun. 

```

### CODE REVIEW

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go [hunk 1]
```diff
@@ -95,11 +99,20 @@ func (rcc *ResolvedConditionCheck) ConditionToTaskSpec() *v1alpha1.TaskSpec {
 		t.Inputs = &v1alpha1.Inputs{
 			Params: rcc.Condition.Spec.Params,
 		}
+		convertParamTemplates(&t.Steps[0], rcc.Condition.Spec.Params)
```
Comment (sbwsg) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: Why is this needed? The comment on `convertParamTemplates` explains what it does but I'm still a bit lost on the reason we do it.

Comment (sbwsg) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: oh wait is this because the condition spec places the params at a higher level than the `ApplyContainerReplacements` func is expecting and we're trying to reuse that logic from taskruns?  might be worth adding a comment to that effect here.

File: pkg/templating/container_replacements_test.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/tektoncd/pipeline/pkg/templating"
+)
+
+func TestApplyContainerReplacements(t *testing.T) {
```
Comment (sbwsg) [inline pkg/templating/container_replacements_test.go]: nice

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go (hunk 1 above)
Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: added!

File: docs/conditions.md [hunk 3]
```diff
@@ -32,11 +33,34 @@ following fields:
...
-runs till completion. The container must exit successfully i.e. with an exit code 0 for the 
-condition evaluation to be successful. All other exit codes are considered to be a condition check
+check must specify a container image that adheres to the [container contract](./container-contract.md). 
+The container image runs till completion. The container must exit successfully i.e. with an exit code 0 
+for the condition evaluation to be successful. All other exit codes are considered to be a condition check
 failure.
 
+### Parameters
+
+A Condition can declare parameters that must be supplied to the it during a PipelineRun. Sub-fields
```
Comment (sbwsg) [inline docs/conditions.md]: s/the it/it/

File: docs/conditions.md [hunk 4]
```diff
@@ -32,11 +33,34 @@ following fields:
...
+  check:
+    image: ${params.image}
+```
+
+Parameters name are limited to alpha-numeric characters, `-` and `_` and can
+only start with alpha characters and `_`. For example, `fooIs-Bar_` is a valid
+parameter name, `barIsBa$` or `0banana` are not.
+ 
+Each declared parameter has a type field, assumed to be string if not provided by the user. 
+The other possible type is array — useful, for instance, when a dynamic number of compilation
```
Comment (sbwsg) [inline docs/conditions.md]: > compilation flags

Would a condition container typically be responsible for compiling something? If not, is there another example that might be more relevant to condition containers?  Maybe something like "checking a pushed branch name doesn't match any of multiple protected branch names" where the protected branch names are passed in as an array param?

File: pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go [hunk 5]
```diff
@@ -95,11 +99,22 @@ func (rcc *ResolvedConditionCheck) ConditionToTaskSpec() *v1alpha1.TaskSpec {
 		t.Inputs = &v1alpha1.Inputs{
 			Params: rcc.Condition.Spec.Params,
 		}
+		// convert param strings of type ${params.x} to ${inputs.params.x}
+		// in order to apply apply taskrun substitution
```
Comment (sbwsg) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: s/apply apply/apply/

Comment (dibyom) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/conditionresolution.go]: :man_facepalming:

File: docs/conditions.md (hunk 4 above)
Comment (dibyom) [reply docs/conditions.md]: great idea! and changed!

**Comment by afrittoli** [comment]
I've not reviewed the code yet, but I noticed that the params are limited to string - I wonder if we should support the newly added array params too? That could be a separate PR of course.

**Comment by dibyom** [comment]
> I wonder if we should support the newly added array params too?

Array params should be supported for the `args` and `command` fields just like in a task. Could not think of a good use case for it for the examples at the moment, but I could add it in!

File: pkg/templating/container_replacements_test.go (hunk 2 above)
Comment (afrittoli) [reply pkg/templating/container_replacements_test.go]: Thanks for improving the test coverage here!
Since you're at it 👼 it would be nice to test the case of replacement when the variable is not defined.
